control-flow/index.md

```diff
@@ -1,6 +1,368 @@
+
+```
+function randomDelayedResponse(text) {
+  // Using a timeout here for the sake of simulating an external request
+  const timeOut = Math.floor(Math.random() * 10) + 1;
+  const output = text;
+  setTimeout(() => {
+    return output;
+  }, timeOut * 10);
+}
+
+const output = randomDelayedResponse('Hello');
+console.log(output); // empty
+```
+
+Notice that the final line above returns `undefined`.
+
+We need to wait until the response is ready. One way is to pass our `console.log` in to `randomDelayedResponse` as a function.
+
+```
+function randomDelayedResponse(text, callback) {
+  // Using a timeout here for the sake of simulating an external request
+  const timeOut = Math.floor(Math.random() * 10) + 1;
+  setTimeout(() => {
+    callback(text);
+  }, timeOut * 10);
+}
+
+const output = randomDelayedResponse('Hello', text => console.log(text)); // outputs "Hello"
+console.log(output); // still empty
+```
+
+We pass the function `text => console.log(text)` as the second parameter, and this function is then called after the `setTimeout`.
+
+This might be reasonable in simple situations but it's easy for callbacks to get out of control. For example, if we have many different calls to this slow function that need to be run in order. To achieve this control we need to nest further.
+
+To show many responses, consider the following code. If we ran them beside each other, the output would not be reliable.
+
+```
+function randomDelayedResponse(text, callback) {
+  // Using a timeout here for the sake of simulating an external request
+  const timeOut = Math.floor(Math.random() * 10) + 1;
+  setTimeout(() => {
+    callback(text);
+  }, timeOut * 10);
+}
+
+randomDelayedResponse(1, text => console.log(text)); // outputs 1
+randomDelayedResponse(2, text => console.log(text)); // outputs 2
+randomDelayedResponse(3, text => console.log(text)); // outputs 3
+randomDelayedResponse(4, text => console.log(text)); // outputs 4
+// Who will win?
+```
+
+We can run the above code multiple times, and the result is unpredictable. It is _asynchronous_, in that the results do not arrive in order. To create a predictable, _synchronous_ flow using callbacks we'd need to nest them.
+
+```
+function randomDelayedResponse(text, callback) {
+  // Using a timeout here for the sake of simulating an external request
+  const timeOut = Math.floor(Math.random() * 10) + 1;
+  setTimeout(() => {
+    callback(text);
+  }, timeOut * 10);
+}
+
+randomDelayedResponse(1, text => {
+  console.log(text);
+  randomDelayedResponse(2, text => {
+    console.log(text);
+    randomDelayedResponse(3, text => {
+      console.log(text);
+      randomDelayedResponse(4, text => {
+        console.log(text);
+      });
+    });
+  });
+}); // outputs "1 2 3 4"
+```
+
+Structuring these callbacks works. However this approach can create code becomes difficult to understand and maintain. Let's look at another way.
```

Comment:
`approach can create code becomes difficult` => `approach can create code that becomes difficult`